File: packages/lynx/ios/Sources/HotUpdaterLynxArtifact/LynxBridgeReplies.swift
import Foundation

/** Settles accepted bridge operations when their owning generation is retired. */
final class LynxBridgeReplies {
    struct Ticket: Hashable { fileprivate let id: Int }

    private let lock = NSLock()
    private var callbacks: [Int: (Error) -> Void] = [:]
    private var nextID = 0
    private var closed = false

    func register(_ cancellation: @escaping (Error) -> Void) -> Ticket? {
        lock.lock()
        guard !closed else {
            lock.unlock()
            cancellation(Self.contextRejected)
            return nil
        }
        let ticket = Ticket(id: nextID)
        nextID += 1
        callbacks[ticket.id] = cancellation
        lock.unlock()
        return ticket
    }

    func settle(_ ticket: Ticket, _ completion: () -> Void) {
        lock.lock()
        let accepted = callbacks.removeValue(forKey: ticket.id) != nil
        lock.unlock()
        if accepted { completion() }
    }

    /** Transfers an accepted reply to a host operation that survives retirement. */
    func claim(_ ticket: Ticket) -> Bool {
        lock.lock()
        let accepted = callbacks.removeValue(forKey: ticket.id) != nil
        lock.unlock()
        return accepted
    }

    func close() {
        lock.lock()
        guard !closed else {
            lock.unlock()
            return
        }
        closed = true
        let pending = Array(callbacks.values)
        callbacks.removeAll()
        lock.unlock()
        pending.forEach { $0(Self.contextRejected) }
    }

    private static let contextRejected = LynxPolicyError(
        code: "CONTEXT_REJECTED",
        message: "The native Lynx context was retired before the operation completed"
    )
}

/** Protects a host-owned asynchronous reply from duplicate completion. */
final class LynxOnceReply<Value> {
    private let lock = NSLock()
    private var callback: ((Result<Value, Error>) -> Void)?

    init(_ callback: @escaping (Result<Value, Error>) -> Void) {
        self.callback = callback
    }

    func settle(_ result: Result<Value, Error>) {
        lock.lock()
        let callback = self.callback
        self.callback = nil
        lock.unlock()
        callback?(result)
    }
}

File: packages/lynx/android/src/main/java/com/hotupdater/lynx/LynxBridgeReplies.kt
package com.hotupdater.lynx

import java.util.concurrent.atomic.AtomicBoolean
import org.json.JSONObject

/** A stable native error code that can cross the Lynx callback envelope. */
class LynxNativeOperationException(
    val code: String,
    message: String,
    cause: Throwable? = null,
) : IllegalStateException(message, cause)

/** Settles accepted bridge operations when their owning generation is retired. */
internal class LynxBridgeReplies {
    internal data class Ticket(val id: Long)

    private val lock = Any()
    private val callbacks = linkedMapOf<Long, (Result<JSONObject>) -> Unit>()
    private var nextId = 0L
    private var closed = false

    fun register(callback: (Result<JSONObject>) -> Unit): Ticket? {
        val ticket = synchronized(lock) {
            if (closed) null else Ticket(nextId++).also {
                callbacks[it.id] = callback
            }
        }
        if (ticket == null) callback(Result.failure(contextRejected()))
        return ticket
    }

    fun settle(ticket: Ticket, result: Result<JSONObject>) {
        val callback = synchronized(lock) { callbacks.remove(ticket.id) }
        callback?.invoke(result)
    }

    fun close() {
        val pending = synchronized(lock) {
            if (closed) return
            closed = true
            callbacks.values.toList().also { callbacks.clear() }
        }
        pending.forEach { it(Result.failure(contextRejected())) }
    }

    private fun contextRejected() = LynxNativeOperationException(
        "CONTEXT_REJECTED",
        "The native Lynx context was retired before the operation completed",
    )
}

/** Protects a host-owned asynchronous reply from duplicate completion. */
internal class LynxOnceReply<T>(
    private val callback: (Result<T>) -> Unit,
) {
    private val settled = AtomicBoolean()

    fun settle(result: Result<T>) {
        if (settled.compareAndSet(false, true)) callback(result)
    }
}
